external/js/cothority/src/network/connection.ts

```diff
@@ -173,35 +178,43 @@ export class RosterWSConnection {
      */
     async send<T extends Message>(message: Message, reply: typeof Message): Promise<T> {
         const errors: string[] = [];
-        let rotate = this.addresses.length - this.connectionsActive.length;
+        const msgNbr = this.msgNbr;
+        this.msgNbr++;
+        const list = this.nodes.newList(this.service);
+        const pool = list.active;
+
+        Log.lvl3(sprintf("%d/%d", this.connNbr, msgNbr), "sending", message.constructor.name, "with list:",
+            pool.map((conn) => conn.getURL()));
 
         // Get the first reply - need to take care not to return a reject too soon, else
         // all other promises will be ignored.
         // The promises that never 'resolve' or 'reject' will later be collected by GC:
         // https://stackoverflow.com/questions/36734900/what-happens-if-we-dont-resolve-or-reject-the-promise
-        return Promise.race(this.connectionsActive.map((_, i) => {
+        return Promise.race(pool.map((conn) => {
             return new Promise<T>(async (resolve, reject) => {
                 do {
+                    const idStr = sprintf("%d/%d: %s - ", this.connNbr, msgNbr.toString(), conn.getURL());
                     try {
-                        const sub = await this.connectionsActive[i].send(message, reply);
-                        // Signal to other connections that have an error that they don't need
-                        // to retry.
-                        rotate = -1;
-                        resolve(sub as T);
+                        Log.lvl3(idStr, "sending");
+                        const sub = await conn.send(message, reply);
+                        Log.lvl3(idStr, "received OK");
+
+                        if (list.done(conn) === 0) {
+                            Log.lvl3(idStr, "first to receive");
+                            resolve(sub as T);
+                        }
+                        return;
                     } catch (e) {
+                        Log.lvl3(idStr, "has error", e);
                         errors.push(e);
-                        if (errors.length === this.addresses.length) {
+                        conn = list.replace(conn);
+                        if (errors.length >= list.length / 2) {
                             // It's the last connection that also threw an error, so let's quit
                             reject(errors);
-                        }
-                        rotate--;
-                        if (rotate >= 0) {
-                            // Take the oldest connection that hasn't been used yet
-                            this.connectionsPool.push(this.connectionsActive[i]);
-                            this.connectionsActive[i] = this.connectionsPool.shift();
+                            conn = undefined;
```

Comment:
This is never excuted. Maybe also `throw` the list instead of `rejecting` it, as it looks clearer that the control flow is changed.